### PR TITLE
JAMES-2586 Fix [PGSQL] Concurrency control for flags updates

### PR DESCRIPTION
**Pull request by vttranlina**
## The before way:
- James will fetch current flags from the database -> evaluate the new updatedFlag based on request -> update the new flags to the database 

## The after way
- Based on the update flags mode (append, remove, replace) then will directly update to flags in the database (without fetching before)

### CODE REVIEW

**Comment by chibenwa** [comment]
So... Does it work?

**Comment by chibenwa** [comment]
And measure performance.

**Comment by vttranlina** [comment]
> So... Does it work?

yes, the concurrent test passed

> And measure performance.

let me

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageMapper.java [hunk 1]
```diff
@@ -321,27 +321,40 @@ private Flux<UpdatedFlags> updatedFlags(List<ComposedMessageIdWithMetaData> list
     }
 
     private Mono<UpdatedFlags> updateFlags(ComposedMessageIdWithMetaData currentMetaData,
-                                           FlagsUpdateCalculator flagsUpdateCalculator,
-                                           ModSeq newModSeq) {
+                                            FlagsUpdateCalculator flagsUpdateCalculator,
+                                            ModSeq newModSeq) {
```
Comment (quantranhong1999) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageMapper.java:325]: The old indent is better?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java [hunk 2]
```diff
@@ -313,46 +317,93 @@ public Flux<ComposedMessageIdWithMetaData> findAllRecentMessageMetadata(Postgres
...
+
+        return updateStatement.get()
+            .set(USER_FLAGS, newFlags.getUserFlags())
+            .set(MOD_SEQ, newModSeq.asLong())
             .where(MAILBOX_ID.eq(mailboxId.asUuid()))
             .and(MESSAGE_UID.eq(uid.asLong()));
     }
 
+    private UpdateConditionStep<Record> buildRemoveFlagsStatement(DSLContext dslContext, Flags removeFlags,
+                                                               PostgresMailboxId mailboxId, MessageUid uid, ModSeq newModSeq) {
```
Comment (quantranhong1999) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:371]: weird indent?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java [hunk 3]
```diff
@@ -313,46 +317,93 @@ public Flux<ComposedMessageIdWithMetaData> findAllRecentMessageMetadata(Postgres
...
-            .set(USER_FLAGS, updatedFlags.getNewFlags().getUserFlags())
-            .set(MOD_SEQ, updatedFlags.getModSeq().asLong())
+            .set(USER_FLAGS, PostgresDSL.arrayCat(USER_FLAGS, addFlags.getUserFlags()))
+            .set(MOD_SEQ, newModSeq.asLong())
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_UID.eq(uid.asLong()));
+    }
+
+    private UpdateConditionStep<Record> buildReplaceFlagsStatement(DSLContext dslContext, Flags newFlags,
+                                                               PostgresMailboxId mailboxId, MessageUid uid, ModSeq newModSeq) {
```
Comment (quantranhong1999) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:356]: weird indent?

**Review by Arsnael — COMMENTED** [review]
Nothing to add. Please rebase as well :)

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageModule.java [hunk 4]
```diff
@@ -108,6 +108,24 @@ interface MessageToMailboxTable {
         Field<String[]> USER_FLAGS = DSL.field("user_flags", DataTypes.STRING_ARRAY);
         Field<LocalDateTime> SAVE_DATE = DSL.field("save_date", DataTypes.TIMESTAMP);
 
+        String ARRAY_REMOVE_JAMES_FUNCTION_NAME = "array_remove_james";
+        String CREATE_ARRAY_REMOVE_JAMES_FUNCTION =
+            "CREATE OR REPLACE FUNCTION " + ARRAY_REMOVE_JAMES_FUNCTION_NAME + "(\n" +
+                "    p_remove_flags_1 text[],\n" +
+                "    p_remove_flags_2 text[])\n" +
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageModule.java:115]: Can you give me a new name?

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageModule.java:115]: ChatGPT suggestion :p
image

BTW IMO we should not put `flags` keyword in this function. Maybe `source` and `elements_to_remove`?

**Review by hungphan227 — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by vttranlina** [comment]
The first performance test looks not good. 
The KO result is better, but the 99th is not.

## Env
Local.
Sample: `select count(*) from message_mailbox;` = 566904
Before the test, I created a sample user_flags by query
```sql
UPDATE message_mailbox
SET user_flags = (
    SELECT ARRAY(SELECT 'sample_data_' || generate_series(1, 50)::text)
) where 1=1
returning user_flags;
```
I also modify the Gatling scenario for increase the update flags request. 
You can see: `singleFlagsAdd`, `rangeFlagsAdd`, `singleFlagsRemove`, `rangeFlagsRemove` 
https://github.com/linagora/james-gatling/commit/be38ae6730e1e4df085e5ef300de2da65dd169d1

Gatling
```
export DURATION=20
export USER_COUNT=5000
```

Result:
updateFlagConcurrency

**Review by chibenwa — COMMENTED** [review]
The performance impact is not neglictible. IMO this is a blocker to merge it as is.

We only need to call the complex function if we update several flags at once correct?

I propose to continue iterating other that code and actually call the PG-function if truely needed, and rely on simpler queries the rest of the time.

WDYT?

**Comment by vttranlina** [comment]
> We only need to call the complex function if we update several flags at once correct?

The point is not "several flags" or "single flag", even when we add "one" flag with the old way, it still does not ensure concurrence,

**Comment by chibenwa** [comment]
Sure read then write would not be concurrent friendly...

> Based on the update flags mode (append, remove, replace) then will directly update to flags in the database (without fetching before)

Even if we use `add_item` ?

**Comment by vttranlina** [comment]
I updated the logic update `user_flags` 
when `1` flag -> array_append, array_remove
when `>1 flag` -> array_cat, array_remove_james

TODO
- [ ] performance test 

// after a lot of performance test local, now it throws a lot of KO result even on `postgresql` branch, I will continue it tomorrow

**Comment by chibenwa** [comment]
I confirm this is what I had in mind.

I assume perf test will go better ;-)

**Review by Arsnael — APPROVED** [review]
You seem to say it's getting worse locally... Maybe not stable enough? Don't hesitate to use charge-01 if it makes a difference :)

**Comment by vttranlina** [comment]
> I assume perf test will go better ;-)

Yes, at this time, it looks better. 
I decreased the sample to: (decrease the waiting time) 
```
export DURATION=10
export USER_COUNT=2500
```

updateFlagConcurrency_2500

**Detail:**
before: https://vttranlina.github.io/gatling_20231220/before_2500/index.html
after: https://vttranlina.github.io/gatling_20231220/after_2500/index.html

**Comment by vttranlina** [comment]
> You seem to say it's getting worse locally... Maybe not stable enough?

I guess because after several tests, the sample data is increased after each run, and when it gets the threshold, it throws
After flushing all data and resetting, it re-normal
The throw related that I got: https://github.com/linagora/james-project/issues/4997

**Review by chibenwa — APPROVED** [review]

**Pull request merged**